google/cloud/spanner/sql_partition.h

```diff
@@ -0,0 +1,131 @@
+ * slice of a parallel SQL read.
+ *
+ * Instances of `SqlPartition` are created by `Client::PartitionSql`. Once
+ * created, `SqlPartition` objects can be serialized, transmitted to separate
+ * process, and used to read data in parallel using `Client::ExecuteSql`.
+ */
+class SqlPartition {
+ public:
+  /**
+   * Constructs an instance of `SqlPartition` that is not associated with any
+   * `SqlStatement`.
+   */
+  SqlPartition() = default;
+
+  // Copy and move.
+  SqlPartition(SqlPartition const&) = default;
+  SqlPartition(SqlPartition&&) = default;
+  SqlPartition& operator=(SqlPartition const&) = default;
+  SqlPartition& operator=(SqlPartition&&) = default;
+
+  /**
+   * Accessor for the `SqlStatement` associated with this `SqlPartition`.
+   * @return SqlStatement
+   */
+  SqlStatement const& sql_statement() const;
+
+ private:
+  friend class SqlPartitionTester;
+  friend SqlPartition internal::MakeSqlPartition(
+      std::string const& transaction_id, std::string const& session_id,
+      std::string const& partition_token, SqlStatement const& sql_statement);
+  friend std::string SerializeSqlPartition(SqlPartition const& sql_partition);
+  friend google::cloud::StatusOr<SqlPartition> DeserializeSqlPartition(
+      std::string const& serialized_sql_partition);
+
+  explicit SqlPartition(std::string transaction_id, std::string session_id,
+                        std::string partition_token,
+                        SqlStatement sql_statement);
+
+  // Accessor methods for use by friends.
+  std::string const& partition_token() const;
+  std::string const& session_id() const;
+  std::string const& transaction_id() const;
+
+  std::string transaction_id_;
+  std::string session_id_;
+  std::string partition_token_;
+  SqlStatement sql_statement_;
+};
+
+/**
+ * Serializes an instance of `SqlPartition` for transmission to another process.
+ * @param sql_partition - instance to be serialized.
+ * @return `std::string`
+ *
+ * @par Example:
+ *
+ * @code
+ * spanner::SqlStatement stmt("select * from Albums");
+ * std::vector<spanner::SqlPartition> partitions =
+ *   spanner_client.PartitionSql(stmt);
+ * for (auto const& partition : partitions) {
+ *   SendToRemoteMachine(spanner::SerializeSqlPartition(partition));
+ * }
+ * @endcode
+ */
+std::string SerializeSqlPartition(SqlPartition const& sql_partition);
+
+/**
+ * Deserialized the provided string into a `SqlPartition`, if able.
+ *
+ * Returned `Status` should be checked to determine if deserialization was
+ * successful.
+ *
+ * @param serialized_sql_partition
+ * @return `google::cloud::StatusOr<SqlPartition>`
+ *
+ * @par Example:
+ *
+ * @code
+ * std::string serialized_partition = ReceiveFromRemoteMachine();
+ * spanner::SqlPartition partition =
+ *   spanner::DeserializeSqlPartition(serialized_partition);
+ * auto rows = spanner_client.ExecuteSql(partition);
+ * @endcode
+ */
+google::cloud::StatusOr<SqlPartition> DeserializeSqlPartition(
```

Comment:
I'm not sure what the "right" answer is, but ftr, in other code I've been spelling this just `StatusOr<Foo>`, and omitting the "google::cloud::" prefix because it's this code's same namespace. I'm not sure if we have a policy or consistency argument in some other direction though.

Reply:
Done.